hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/placement/algorithms/SCMContainerPlacementRackScatter.java

```diff
@@ -0,0 +1,343 @@
+            " data size required: " + dataSizeRequired;
+        LOG.info(errMsg);
+        return null;
+      }
+      if (excludedNodesForCapacity == null) {
+        excludedNodesForCapacity = new ArrayList<>();
+      }
+      excludedNodesForCapacity.add(node.getNetworkFullPath());
+    }
+  }
+
+  /**
+   * This default implementation handles rack aware policies and non rack
+   * aware policies. If a future placement policy needs to check more than racks
+   * to validate the policy (eg node groups, HDFS like upgrade domain) this
+   * method should be overridden in the sub class.
+   * This method requires that subclasses which implement rack aware policies
+   * override the default method getRequiredRackCount and getNetworkTopology.
+   * @param dns List of datanodes holding a replica of the container
+   * @param replicas The expected number of replicas
+   * @return ContainerPlacementStatus indicating if the placement policy is
+   *         met or not. Not this only considers the rack count and not the
+   *         number of replicas.
+   */
+  @Override
+  public ContainerPlacementStatus validateContainerPlacement(
+      List<DatanodeDetails> dns, int replicas) {
+    int requiredRacks = replicas;
+    if (networkTopology == null || replicas == 1 || requiredRacks == 1) {
+      return super.validateContainerPlacement(dns, replicas);
+    }
+    // We have a network topology so calculate if it is satisfied or not.
+    int numRacks = 1;
+    final int maxLevel = networkTopology.getMaxLevel();
+    // The leaf nodes are all at max level, so the number of nodes at
+    // leafLevel - 1 is the rack count
+    numRacks = networkTopology.getNumOfNodes(maxLevel - 1);
+    final long currentRackCount = dns.stream()
+        .map(d -> networkTopology.getAncestor(d, 1))
+        .distinct()
+        .count();
+
+    if (replicas > numRacks) {
+      requiredRacks = numRacks;
+    }
+    return new ContainerPlacementStatusDefault(
+        (int)currentRackCount, requiredRacks, numRacks);
+  }
+
+  private Node getRackOfDatanodeDetails(DatanodeDetails datanodeDetails) {
+    String location = datanodeDetails.getNetworkLocation();
+    return networkTopology.getNode(location);
+  }
+
+  private List<Node> sortRackWithExcludedNodes(List<Node> racks,
+      List<DatanodeDetails> excludedNodes) {
+    Set<Node> lessPreferedRacks = excludedNodes.stream()
+        .map(node -> networkTopology.getAncestor(node, RACK_LEVEL))
+        .collect(Collectors.toSet());
+    List <Node> result = new ArrayList<>();
+    for (Node rack : racks) {
+      if (!lessPreferedRacks.contains(rack)) {
+        result.add(rack);
+      }
+    }
+    result.addAll(lessPreferedRacks);
+    return result;
+  }
+
+  private List<Node> getAllRacks() {
+    int rackLevel = networkTopology.getMaxLevel()-1;
+    return networkTopology.getNodes(rackLevel);
+  }
+
+  private boolean isValidNode(DatanodeDetails datanodeDetails,
```

Comment:
The logic here is very similar to that in `SCMContainerPlacementRackAware.chooseNode()` - do you think we could add this method (`isValidNode()`) to `SCMCommonPlacmentPolicy` and also refactor the class above to use it too?

Reply:
Will add these changes to master.